<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/2_transfer_learning_roadmap/2_acceptable_data_types/3)%20Multi%20label%20-%20Data%20with%20labels%20in%20CSV%20file.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals

### 1. Understand what multi label image classifications is
        - One or more than one label per image

### 2. Understand multi label - csv type data structure

## Table of Contents


## [0. Install](#0)


## [1. Load train data - csv type - Multi Label](#1)


## [2. Understand experiment data-summary](#2)


## [3. Load val data along with train data](#3)


## [4. Understand experiment data-summary](#4)

# Dataset Details
- Credits: https://www.kaggle.com/c/planet-understanding-the-amazon-from-space

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=12ia20erQXTmJTk1zj1pZ_VeCPViyAQvm' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=12ia20erQXTmJTk1zj1pZ_VeCPViyAQvm" -O satellite_cls.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq satellite_cls.zip

<a id='0'></a>
# Install Monk
  
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 - cd monk_v1/installation/Linux && pip install -r requirements_cu9.txt
     - (Select the requirements file as per OS and CUDA version)

In [3]:
!git clone https://github.com/Tessellate-Imaging/monk_v1.git

Cloning into 'monk_v1'...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 1916 (delta 4), reused 10 (delta 4), pack-reused 1905
Receiving objects: 100% (1916/1916), 69.95 MiB | 10.27 MiB/s, done.
Resolving deltas: 100% (1024/1024), done.


In [ ]:
# If using Colab install using the commands below
!cd monk_v1/installation/Misc && pip install -r requirements_colab.txt

# If using Kaggle uncomment the following command
#!cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt

# Select the requirements file as per OS and CUDA version when using a local system or cloud
#!cd monk_v1/installation/Linux && pip install -r requirements_cu9.txt

# Imports

In [1]:
# Monk
import os
import sys
sys.path.append("monk_v1/monk/");

In [2]:
#Using mxnet-gluon backend 
from gluon_prototype import prototype

## Creating and managing experiments
    - Provide project name
    - Provide experiment name

In [7]:
gtf = prototype(verbose=1);
gtf.Prototype("Dataset-Type", "exp-csv-multi-label");

Mxnet Version: 1.5.0

Experiment Details
    Project: Dataset-Type
    Experiment: exp-csv-multi-label
    Dir: /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.4_multilabel/workspace/Dataset-Type/exp-csv-multi-label/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Dataset-Type 
                        |
                        |
                        |-----exp-foldered 
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

<a id='1'></a>
# Load train data - foldered type

## Quick mode training

    - Using Default Function
        - dataset_path
        - model_name
        - num_epochs
        
        
## Dataset folder structure

    parent_directory
        |
        |
        |------train (folder can be named anything)
                |------img1.jpg
                |------img2.jpg
                |------.... (and so on)
        |
        |------train_labels.csv (file can be named anything)
        
        
    train_labels.csv has 2 columns (column names could be anything)
        - ID                 | Labels
          img1.jpg           |  agriculture cloudy
          img2.jpg           |  forest hazy
          img3.jpg           |  road forest rains
          img4.jpg           |  city
                ..... (and so on)
                
### In similar fashion you can structure your own dataset

### Along with providing dataset_path you now provide path_to_csv

In [8]:
gtf.Default(dataset_path="satellite_cls/train-jpg",
            path_to_csv="satellite_cls/train.csv",
            model_name="resnet18_v1", 
            
            delimiter = " ",    # very important to mention how the labels are separated, via spaces, or commas, etc
            
            
            num_epochs=5);

Dataset Details
    Train path:     satellite_cls/train-jpg
    Val path:       None
    CSV train path: satellite_cls/train.csv
    CSV val path:   None
    Label Type:     multiple

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   4
    Train-val split:   0.7
    Delimiter:    

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 28335
    Num val images:   12144
    Num classes:      17

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
    

<a id='2'></a>
# Understand experiment data-summary

    Dataset Details
        Train path:     monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train
        Val path:       None
        CSV train path: monk_v1/monk/system_check_tests/datasets/dataset_csv_id/train.csv
        CSV val path:   None
        
        
### Train path is mentioned and Val path is None

### CSV Train path is mentioned unlike in foldered type dataset


    Dataset Parameters:
        Input Size:   224
        Batch Size:   4
        Shuffle:      True
        Processors:   4
        Num Classes:  2

###  Data input size is 224 (default setting)
###  Batch size is 4 (default setting)

   Dataset Numbers
    Num train images: 28335
    Num val images:   12144
    Num classes:      17
        
### Since val data is missing, train data is split into train and val with a ratio 70:30 rule
    - 70% of data being used for training
    - 30% of data being used for validation
    
### Auto detect number of classes